cmd/cosign/cli/trustedroot/trustedroot.go

```diff
@@ -0,0 +1,211 @@
+//
+// Copyright 2024 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trustedroot
+
+import (
+	"context"
+	"crypto"
+	"crypto/x509"
+	"encoding/hex"
+	"encoding/pem"
+	"fmt"
+	"os"
+
+	"github.com/sigstore/sigstore-go/pkg/root"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+
+	"github.com/sigstore/cosign/v2/pkg/cosign"
+)
+
+type CreateCmd struct {
+	CAIntermediates  string
+	CARoots          string
+	CertChain        string
+	CtfeKeyPath      string
+	Out              string
+	RekorKeyPath     string
+	TSACertChainPath string
+}
+
+func (c *CreateCmd) Exec(_ context.Context) error {
+	var fulcioCertAuthorities []root.CertificateAuthority
+	ctLogs := make(map[string]*root.TransparencyLog)
+	var timestampAuthorities []root.CertificateAuthority
+	rekorTransparencyLogs := make(map[string]*root.TransparencyLog)
+
+	if c.CertChain != "" {
+		fulcioAuthority, err := parsePEMFile(c.CertChain)
+		if err != nil {
+			return err
+		}
+		fulcioCertAuthorities = append(fulcioCertAuthorities, *fulcioAuthority)
+	} else if c.CARoots != "" {
+		roots, err := parseCerts(c.CARoots)
+		if err != nil {
+			return err
+		}
+
+		var intermediates []*x509.Certificate
+		if c.CAIntermediates != "" {
+			intermediates, err = parseCerts(c.CAIntermediates)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Here we're trying to "flatten" the x509.CertPool cosign was using
+		// into a trusted root with a clear mapping between roots and
+		// intermediates. Make a guess that if there are intermediates, there
+		// is one per root.
+
+		for i, rootCert := range roots {
+			var fulcioAuthority root.CertificateAuthority
+			fulcioAuthority.Root = rootCert
+			if i < len(intermediates) {
+				fulcioAuthority.Intermediates = []*x509.Certificate{intermediates[i]}
+			}
+			fulcioCertAuthorities = append(fulcioCertAuthorities, fulcioAuthority)
+		}
+	}
+
+	if c.CtfeKeyPath != "" {
+		ctLogPubKey, id, idBytes, err := getPubKey(c.CtfeKeyPath)
+		if err != nil {
+			return err
+		}
+
+		ctLogs[id] = &root.TransparencyLog{
```

Comment:
Do we need to set validity periods as well? Do we need to take in flags for start time, or default to now?

Reply:
Since we aren't setting `ValidityPeriodStart`, it will default to "0001-01-01T00:00:00Z".

Philisophically, I was thinking of this utility more as "enough of a trusted root to get you started" instead of a flexible tool to cover all use-cases. This will be a recurring theme in my responses! And of course once you have the skeleton of a trusted root file, the user can continue editing it and adding additional content.

Reply:
Agreed it shouldn't cover all cases, but if we're going to instruct users to go manually inspect it afterwards and update a set of values, providing flags to do that for them seems reasonable. Since validity period is a significant benefit of the trusted root, I'd prefer adding a flag for users to set the start time.

Reply:
I added the ability to specify the validity start time for keys